src/Dfc.CourseDirectory.WebV2/Features/Providers/EditProviderType.cs

```diff
@@ -40,25 +52,77 @@ public async Task<Command> Handle(Query request, CancellationToken cancellationT
+            var providerTLevelDefinitions = provider.ProviderType.HasFlag(ProviderType.TLevels)
+                ? await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitionsForProvider { ProviderId = request.ProviderId })
+                : Enumerable.Empty<SqlModels.TLevelDefinition>();
+
             return new Command()
             {
                 ProviderId = provider.Id,
-                ProviderType = provider.ProviderType
+                ProviderType = provider.ProviderType,
+                ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, providerTLevelDefinitions.Select(pd => pd.TLevelDefinitionId))
             };
         }
 
-        public async Task<Success> Handle(Command request, CancellationToken cancellationToken)
+        public async Task<OneOf<ModelWithErrors<Command>, Success>> Handle(Command request, CancellationToken cancellationToken)
         {
+            var tLevelDefinitions = await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitions());
+
+            var validationResult = await new CommandValidator(tLevelDefinitions).ValidateAsync(request);
+
+            if (!validationResult.IsValid)
+            {
+                request.ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, request.SelectedProviderTLevelDefinitionIds);
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
             await _cosmosDbQueryDispatcher.ExecuteQuery(new UpdateProviderType()
             {
                 ProviderId = request.ProviderId,
                 ProviderType = request.ProviderType
             });
 
+            await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.UpsertProviderTLevelDefinitions
+            {
+                ProviderId = request.ProviderId,
+                TLevelDefinitionIds = request.ProviderType.HasFlag(ProviderType.TLevels)
+                    ? request.SelectedProviderTLevelDefinitionIds ?? Enumerable.Empty<Guid>()
+                    : Enumerable.Empty<Guid>()
+            });
+
             // Remove this provider from the cache - subsequent requests will re-fetch updated record
             await _providerInfoCache.Remove(request.ProviderId);
 
             return new Success();
         }
+
+        private class CommandValidator : AbstractValidator<Command>
+        {
+            public CommandValidator(IEnumerable<SqlModels.TLevelDefinition> tLevelDefinitions)
+            {
+                When(c => c.ProviderType.HasFlag(ProviderType.TLevels), () =>
+                {
+                    RuleFor(c => c.SelectedProviderTLevelDefinitionIds)
+                        .NotEmpty()
+                        .WithMessage("Select the T Levels this provider can offer");
+
+                    When(c => c.SelectedProviderTLevelDefinitionIds?.Any() ?? false, () =>
```

Comment:
I think this needs a message if we're going to add this rule.

Reply:
Alternatively, could remove any invalid entries in `SelectedProviderTLevelDefinitionIds` early on to ignore them.

Reply:
This is a data integrity validation - the message is the default (see unit tests), is adequately descriptive and will not be shown to users under typical failure scenarios. (They'd have to fudge the request or have a data race condition which is unlikely)

Reply:
In other places where we do similar data integrity checks we just throw `InvalidStateException` and let the default error page show. No reason to display a pretty message in these situations, I think. All our current validation messages are curated; this blurs that line.

Reply:
I'd rather keep all of the model validation logic in a single place, however I'm not averse to it throwing an exception. Although I would argue that is a sub-optimal behaviour from the user's point of view given that there may be a change in the underlying data between view render and post that causes the error but is more recoverable with an error message rather than a generic exception page.